flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializerGenerator.java

```diff
@@ -0,0 +1,259 @@
+		}
+		try {
+			serializerClazz = InstantiationUtil.compile(clazz.getClassLoader(), fullClassName, code);
+		}
+		catch (Exception e) {
+			throw new RuntimeException("Unable to generate serializer: " + className, e);
+		}
+		Constructor<?>[] ctors = serializerClazz.getConstructors();
+		assert ctors.length == 1;
+		try {
+			return (TypeSerializer<T>) ctors[0].newInstance(new Object[]{clazz, fieldSerializers, config});
+		} catch (Exception e) {
+			throw new RuntimeException("Unable to instantiate serializer: " + className, e);
+		}
+
+	}
+
+	private void generateCode(String className) {
+		assert fieldSerializers.length > 0;
+		String typeName = clazz.getCanonicalName();
+		StringBuilder members = new StringBuilder();
+		for (int i = 0; i < fieldSerializers.length; ++i) {
+			members.append(String.format("final TypeSerializer f%d;\n", i));
+		}
+		StringBuilder initMembers = new StringBuilder();
+		for (int i = 0; i < fieldSerializers.length; ++i) {
+			initMembers.append(String.format("f%d = serializerFields[%d];\n", i, i));
+		}
+		StringBuilder createFields = new StringBuilder();
+		for (int i = 0; i < fieldSerializers.length; ++i) {
+			createFields.append(String.format("((" + typeName + ")t)." + modifyStringForField(refFields[i],
+				"f%d.createInstance()") + ";\n", i));
+		}
+		StringBuilder copyFields = new StringBuilder();
+		copyFields.append("Object value;\n");
+		for (int i = 0; i < fieldSerializers.length; ++i) {
+			if (refFields[i].getType().isPrimitive()) {
+				copyFields.append(String.format("((" + typeName + ")target)." + modifyStringForField(refFields[i],
+					"f%d.copy(((" + typeName + ")from)." + accessStringForField(refFields[i])) + ");\n", i));
+			} else {
+				copyFields.append(String.format(
+					"value = ((" + typeName + ")from)." + accessStringForField(refFields[i]) + ";\n" +
+					"if (value != null) {\n" +
+					"	((" + typeName + ")target)." + modifyStringForField(refFields[i], "f%d.copy(value)") + ";\n" +
+					"} else {\n" +
+					"	((" + typeName + ")target)." + modifyStringForField(refFields[i], "null") + ";\n" +
+					"}\n", i));
+			}
+		}
+		StringBuilder reuseCopyFields = new StringBuilder();
+		reuseCopyFields.append("Object value;\n");
+		reuseCopyFields.append("Object reuseValue;\n");
+		reuseCopyFields.append("Object copy;\n");
+		for (int i = 0; i < fieldSerializers.length; ++i) {
+			if (refFields[i].getType().isPrimitive()) {
+				reuseCopyFields.append(String.format("((" + typeName + ")reuse)." + modifyStringForField(refFields[i],
+					"f%d.copy(((" + typeName + ")from)." + accessStringForField(refFields[i]) + ")") + ";\n", i));
+			} else {
+				reuseCopyFields.append(String.format(
+					"value = ((" + typeName + ")from)." + accessStringForField(refFields[i]) + ";\n" +
+					"if (value != null) {\n" +
+					"	reuseValue = ((" + typeName + ")reuse)." + accessStringForField(refFields[i]) + ";\n" +
+					"	if (reuseValue != null) {\n" +
```

Comment:
Maybe it's worth it to specialize here for primitive types, since in that case you can omit all these null checks.